cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -159,6 +159,31 @@ public ProgramStatus getProgramStatus(ProgramId programId) throws Exception {
     return getExistingAppProgramStatus(appSpec, programId);
   }
 
+  /**
+   * Returns the program run count of the given program.
+   *
+   * @param programId the id of the program for which the count call is made
+   * @return the run count of the program
+   * @throws NotFoundException if the application to which this program belongs was not found or the program is not
+   *                           found in the app
+   */
+  public int getProgramRuncount(ProgramId programId) throws Exception {
+    AuthorizationUtil.ensureAccess(programId, authorizationEnforcer, authenticationContext.getPrincipal());
+    // check that app exists
+    ApplicationId appId = programId.getParent();
+    ApplicationSpecification appSpec = store.getApplication(appId);
```

Comment:
So these calls on `store` are in different transaction, right? Meaning the check for existence may not be consistent?

Reply:
We have this logic everywhere in this service and many other places. If we want to get rid of that, we will have to use `AppMetadataStore` directly or put this check in the Store method. Which do you think is better? I also think we should spend a lot of time fixing all of these logics in the future